cps/elf.c

```diff
@@ -0,0 +1,266 @@
+		return NULL;
+
+	/* Read in necessary variables from ELF header and fix endianness. */
+	END(hdr->e_shoff, conv, &e_shoff_ntv);
+	END(hdr->e_shnum, conv, &e_shnum_ntv);
+	END(hdr->e_shstrndx, conv, &e_shstrndx_ntv);
+
+	/* Check whether there's enough room for the section headers. */
+	if (len < e_shoff_ntv + e_shnum_ntv * sizeof(sechdrs[0]))
+		return NULL;
+
+	sechdrs = (const ElfPERBIT(Shdr) *)
+		((const uint8_t *)hdr + e_shoff_ntv);
+
+	/* Check whether there's enough room for section with section names. */
+	END(sechdrs[e_shstrndx_ntv].sh_offset, conv, &sh_offset_ntv);
+	if (len < sh_offset_ntv)
+		return NULL;
+
+	/* Find symbol table. */
+	secstrings = (const char *)((const uint8_t *)hdr + sh_offset_ntv);
+
+	for (i = 1; i < e_shnum_ntv; i++) {
+		ElfPERBIT(Word) sh_name_ntv;
+		const char *found_secname;
+
+		END(sechdrs[i].sh_name, conv, &sh_name_ntv);
+		found_secname = secstrings + sh_name_ntv;
+
+		if (strcmp(secname, found_secname) == 0) {
+			ElfPERBIT(Word) sh_size_ntv;
+			END(sechdrs[i].sh_size, conv, &sh_size_ntv);
+			END(sechdrs[i].sh_offset, conv, &sh_offset_ntv);
+
+			/* Not enough room in file for this section. */
+			if (len < sh_offset_ntv + sh_size_ntv)
+				return NULL;
+
+			*secsize = sh_size_ntv;
+			return (const uint8_t *)hdr + sh_offset_ntv;
+		}
+	}
+
+	return NULL;
+}
+
+/* Find the next string in an ELF section. */
+static const char *
+next_string(const char *string, unsigned long *secsize)
+{
+	if (*secsize == 0)
+		return NULL;
+
+	/* Skip non-zero chars. */
+	while (string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	/* Skip any zero padding. */
+	while (!string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	return string;
+}
+
+static int
+load_string(const struct elf_file *mod, const char *attr,
+	char *val, size_t val_len)
+{
+	unsigned long secsize;
+	size_t attr_len;
+	const char *strings;
+
+	/* Get strings from modinfo section. */
+	strings = get_section(mod, ".modinfo", &secsize);
+	if (strings == NULL) {
+		CPS_LOG(ERR, "Unable to get modinfo section from kernel module\n");
+		return -1;
+	}
+
+	if (secsize == 0) {
+		CPS_LOG(ERR, "Returned modinfo section is of length 0\n");
+		return -1;
+	}
+
+	if (strings[secsize - 1] != '\0') {
+		CPS_LOG(ERR, "An unterminated string was found at the end of the modinfo section\n");
+		return -1;
+	}
+
+	/* Skip any zero padding. */
+	while (!strings[0]) {
+		strings++;
+		if (secsize-- <= 1)
+			return -1;
+	}
+
+	attr_len = strlen(attr);
+	while (strings != NULL) {
+		/* Find line with attribute followed by '='. */
+		if (strncmp(strings, attr, attr_len) == 0 &&
+				strings[attr_len] == '=') {
+			if (strlen(strings) - (attr_len + 1) > val_len - 1) {
+				CPS_LOG(ERR, "Found attribute %s in modinfo but value buffer is too short to read in its value (%s)\n",
+					attr, strings);
+				return -1;
+			}
+
+			/* Copy starting from after the '='. */
+			strncpy(val, strings + attr_len + 1, val_len - 1);
```

Comment:
Replace this line with `strcpy(val, strings);`

Notice that the third parameter of `strncpy()` would've needed to be `val_len` not `val_len - 1`.